File: frameworks/ts/src/envelope.ts
/**
 * NDJSON envelope types + parse / write / hash helpers.
 *
 * The envelope protocol matches the Python and Rust frameworks:
 *   data:  {"t":"d","id":"<8B hex>","src":"<string>","v":<payload>}
 *   meta:  {"t":"m","v":<payload>}
 *   log:   {"type":"log","level":"...","msg":"...","...":"..."}  (stderr)
 *   error: {"type":"error","error":"...","input":..,"id":..,"src":..} (stderr)
 */

import { createHash } from 'node:crypto';

export type JSONValue = null | boolean | number | string | JSONValue[] | { [k: string]: JSONValue };

export interface DataEnvelope {
  t: 'd';
  id: string;
  src: string;
  v: JSONValue;
}

export interface MetaEnvelope {
  t: 'm';
  v: JSONValue;
}

export type Envelope = DataEnvelope | MetaEnvelope;

/** blake2b-64 hash of a string. Returns 16-char hex. Matches Python/Rust envelope ID. */
export function hashString(key: string): string {
  return createHash('blake2b512').update(key, 'utf8').digest('hex').slice(0, 16);
}

/** Stream-hash a file by path. Returns 16 hex chars. `undefined` if the file cannot be opened.
 *
 * Streams chunks through the hasher rather than loading the file into a single
 * Uint8Array — important for large inputs (> a few hundred MB).
 */
export async function hashFile(filepath: string): Promise<string | undefined> {
  try {
    const hasher = createHash('blake2b512');
    const stream = Bun.file(filepath).stream();
    for await (const chunk of stream as AsyncIterable<Uint8Array>) {
      hasher.update(chunk);
    }
    return hasher.digest('hex').slice(0, 16);
  } catch {
    return undefined;
  }
}

/** Deterministic id from (src | stage | sorted JSON of v). */
export function computeId(src: string, stage: string, v: JSONValue): string {
  const vStr = canonicalJson(v);
  return hashString(`${src}|${stage}|${vStr}`);
}

/** Parse one NDJSON line. Returns null on blank, invalid JSON, or any
 *  shape that isn't a well-formed envelope. Data envelopes must carry
 *  `id` and `src` strings; meta envelopes must carry a `v` payload. */
export function parseEnvelope(line: string): Envelope | null {
  const trimmed = line.trim();
  if (!trimmed) return null;
  try {
    const obj = JSON.parse(trimmed);
    if (!obj || typeof obj !== 'object') return null;
    if (obj.t === 'd') {
      if (typeof obj.id !== 'string' || typeof obj.src !== 'string') return null;
      if (!('v' in obj)) return null;
      return obj as DataEnvelope;
    }
    if (obj.t === 'm') {
      if (!('v' in obj)) return null;
      return obj as MetaEnvelope;
    }
    return null;
  } catch {
    return null;
  }
}

export function writeData(v: JSONValue, id: string, src: string): void {
  _writeStdout({ t: 'd', id, src, v });
}

export function writeMeta(v: JSONValue): void {
  _writeStdout({ t: 'm', v });
}

export function writeLog(msg: string, level = 'info', extra?: Record<string, JSONValue>): void {
  _writeStderr({ type: 'log', level, msg, ...(extra ?? {}) });
}

export function writeError(v: JSONValue, err: unknown, id: string, src: string): void {
  _writeStderr({
    type: 'error',
    error: err instanceof Error ? err.message : String(err),
    input: v,
    id,
    src,
  });
}

/** Merged trace event flushed once per output envelope (even with empty labels). */
export function writeTrace(id: string, src: string, labels: Record<string, JSONValue>): void {
  _writeStderr({ type: 'trace', id, src, labels });
}

/** Custom stats event. Shape: {type:"stats", ...data}. */
export function writeStats(data: Record<string, JSONValue>): void {
  _writeStderr({ type: 'stats', ...data });
}

/** Canonical JSON for hashing: sorted keys, compact, no spaces. */
function canonicalJson(v: JSONValue): string {
  if (v === null || typeof v !== 'object') return JSON.stringify(v);
  if (Array.isArray(v)) return `[${v.map(canonicalJson).join(',')}]`;
  const keys = Object.keys(v).sort();
  const parts = keys.map(
    (k) => `${JSON.stringify(k)}:${canonicalJson((v as Record<string, JSONValue>)[k]!)}`,
  );
  return `{${parts.join(',')}}`;
}

function _writeStdout(obj: Record<string, JSONValue>): void {
  process.stdout.write(`${JSON.stringify(obj)}\n`);
}

function _writeStderr(obj: Record<string, JSONValue>): void {
  process.stderr.write(`${JSON.stringify(obj)}\n`);
}

File: frameworks/rust/src/envelope.rs
//! NDJSON envelope handling — parse, create, hash utilities.

use blake2::{Blake2b, Digest};
use blake2::digest::consts::U8;
use serde_json::Value;
use std::fs::File;
use std::io::{Read, Write};

type Blake2b64 = Blake2b<U8>;

/// Hash a string. Returns 16-char hex (8 bytes blake2b).
pub fn hash_string(key: &str) -> String {
    let mut hasher = Blake2b64::new();
    hasher.update(key.as_bytes());
    let result = hasher.finalize();
    hex::encode(result)
}

/// Hash file content in chunks. Returns hex string or None on error.
pub fn hash_file(filepath: &str, chunk_size: usize) -> Option<String> {
    let mut file = File::open(filepath).ok()?;
    let mut hasher = Blake2b::<blake2::digest::consts::U16>::new();
    let mut buf = vec![0u8; chunk_size];
    loop {
        let n = file.read(&mut buf).ok()?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    Some(hex::encode(hasher.finalize()))
}

/// Parse a single NDJSON line into envelope. Returns None on parse error.
pub fn parse_envelope(line: &str) -> Option<Value> {
    let trimmed = line.trim();
    if trimmed.is_empty() {
        return None;
    }
    serde_json::from_str(trimmed).ok()
}

/// Write a data envelope to stdout.
pub fn write_data(v: &Value, id: &str, src: &str, out: &mut dyn Write) {
    let envelope = serde_json::json!({
        "t": "d",
        "id": id,
        "src": src,
        "v": v,
    });
    let _ = writeln!(out, "{}", serde_json::to_string(&envelope)
            .expect("envelope serializes — Value tree, no non-string map keys"));
    let _ = out.flush();
}

/// Write a metadata envelope to stdout.
pub fn write_meta(v: &Value, out: &mut dyn Write) {
    let envelope = serde_json::json!({
        "t": "m",
        "v": v,
    });
    let _ = writeln!(out, "{}", serde_json::to_string(&envelope)
            .expect("envelope serializes — Value tree, no non-string map keys"));
    let _ = out.flush();
}

/// Write structured log to stderr.
pub fn write_log(msg: &str, level: &str, err_out: &mut dyn Write) {
    let record = serde_json::json!({
        "type": "log",
        "level": level,
        "msg": msg,
    });
    let _ = writeln!(err_out, "{}", serde_json::to_string(&record)
            .expect("record serializes — Value tree, no non-string map keys"));
    let _ = err_out.flush();
}

/// Write error to stderr with original input preserved.
pub fn write_error(v: &Value, err: &str, id: &str, src: &str, err_out: &mut dyn Write) {
    let record = serde_json::json!({
        "type": "error",
        "error": err,
        "input": v,
        "id": id,
        "src": src,
    });
    let _ = writeln!(err_out, "{}", serde_json::to_string(&record)
            .expect("record serializes — Value tree, no non-string map keys"));
    let _ = err_out.flush();
}

/// Write merged trace event to stderr. Emitted by ctx before each output.
pub fn write_trace(id: &str, src: &str, labels: &Value, err_out: &mut dyn Write) {
    let record = serde_json::json!({
        "type": "trace",
        "id": id,
        "src": src,
        "labels": labels,
    });
    let _ = writeln!(err_out, "{}", serde_json::to_string(&record)
            .expect("record serializes — Value tree, no non-string map keys"));
    let _ = err_out.flush();
}

/// Write a stats event to stderr.
pub fn write_stats(data: &Value, err_out: &mut dyn Write) {
    // data is expected to be a JSON object; merge with type:"stats".
    let mut record = serde_json::json!({ "type": "stats" });
    if let Some(obj) = data.as_object() {
        for (k, v) in obj {
            record[k] = v.clone();
        }
    }
    let _ = writeln!(err_out, "{}", serde_json::to_string(&record)
            .expect("record serializes — Value tree, no non-string map keys"));
    let _ = err_out.flush();
}

// Inline hex encoding to avoid external dependency
mod hex {
    const HEX_CHARS: &[u8; 16] = b"0123456789abcdef";

    pub(super) fn encode(bytes: impl AsRef<[u8]>) -> String {
        let bytes = bytes.as_ref();
        let mut s = String::with_capacity(bytes.len() * 2);
        for &b in bytes {
            s.push(HEX_CHARS[(b >> 4) as usize] as char);
            s.push(HEX_CHARS[(b & 0x0f) as usize] as char);
        }
        s
    }
}
